src/enemy.rs

```diff
@@ -0,0 +1,287 @@
+    let data = serde_json::from_str::<Vec<EnemyConfig>>(&conf).unwrap();
+
+    let mut spawn_map: HashMap<u16, Vec<EnemySpawnComponent>> = HashMap::new();
+
+    for enemy_conf in data {
+        let texture_handle: Handle<Image> = asset_server.load(enemy_conf.asset_path);
+        let mut enemy = EnemySpawnComponent::new(
+            enemy_conf.name,
+            Enemy,
+            Health(enemy_conf.hp),
+            Damage(enemy_conf.dmg),
+            MaxVelocity(enemy_conf.max_velocity),
+            MaxForce(enemy_conf.max_force),
+            Mass(enemy_conf.mass),
+            texture_handle,
+            enemy_conf.is_elite,
+            Rewards { exp: 1, items: "Orange" }, // TODO: Make them drop gems)
+        );
+
+        for waves in enemy_conf.spawn_waves {
+            for n in waves.from..waves.to + 1 {
+                if let Some(mut components) = spawn_map.get_mut(&n) {
+                    enemy.timer = Timer::new(Duration::from_secs(waves.spawn_time), TimerMode::Repeating);
+                    components.push(enemy.clone());
+                } else {
+                    let mut spawn_vec = Vec::new();
+                    enemy.timer = Timer::new(Duration::from_secs(waves.spawn_time), TimerMode::Repeating);
+                    spawn_vec.push(enemy.clone());
+                    spawn_map.insert(n, spawn_vec);
+                }
+            }
+        }
+    }
+
+    let spawners = EnemySpawners(spawn_map);
+    dbg!(spawners.clone());
+    commands.insert_resource(spawners.clone());
+}
+fn spawn(
+    mut commands: Commands,
+    mut spawn_map: ResMut<EnemySpawners>,
+    mut player: Query<&mut Transform, With<Player>>,
+    mut current_wave: ResMut<CurrentWave>,
+    mut global_time_ticker: Res<GlobalTimeTickerResource>,
+) {
+    if global_time_ticker.0.finished() || current_wave.need_wave_spawn {
+        if let Some(spawners) = spawn_map.0.get_mut(&current_wave.num) {
+            for mut spawner in spawners {
+                spawner.timer.tick(global_time_ticker.0.duration());
+
+                if spawner.timer.finished() || current_wave.need_wave_spawn == true {
+                    // spawn elite only one time on the wave
+                    if Some(true) == spawner.is_elite && current_wave.need_wave_spawn == false {
+                        continue
+                    }
+                    if let Ok(p_t) = player.get_single_mut() {
+                        let is_left = thread_rng().gen_range(0, 2);
+                        let is_up = thread_rng().gen_range(0, 2);
+                        let mut enemy_batch: Vec<(EnemyBundle, SpriteBundle, SteeringBundle, Name)> = Vec::new();
+
+                        for i in 1..thread_rng().gen_range(10, 20) {
+                            let (mut m_x, mut m_y, mut m_z): (f32, f32, f32);
+                            if is_left == 1 {
+                                m_x = thread_rng().gen_range(p_t.translation.x - 700.0, p_t.translation.x - 300.0);
```

Comment:
`(p_t.translation.x - 700.0, p_t.translation.x - 300.0)`
We probably need to explicitly name these magical values and move to another struct.

Reply:
done